### ISSUE TITLE
[REVIEW]: Welcome to the tidyverse

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#1686]
**Submitting author:** @hadley (<a href="http://orcid.org/0000-0003-4757-117X">Hadley Wickham</a>)
**Repository:** <a href="http://github.com/tidyverse/tidyverse" target ="_blank">http://github.com/tidyverse/tidyverse</a>
**Version:** v1.3.0
**Editor:** @karthik
**Reviewer:** @ldecicco-USGS, @jeffreyhanson
**Archive:** <a href="https://doi.org/10.5281/zenodo.3547813" target="_blank">10.5281/zenodo.3547813</a>

## Status

[![status](https://joss.theoj.org/papers/3457fa44beec99c5585e1edb601cee61/status.svg)](https://joss.theoj.org/papers/3457fa44beec99c5585e1edb601cee61)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/3457fa44beec99c5585e1edb601cee61"><img src="https://joss.theoj.org/papers/3457fa44beec99c5585e1edb601cee61/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/3457fa44beec99c5585e1edb601cee61/status.svg)](https://joss.theoj.org/papers/3457fa44beec99c5585e1edb601cee61)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="http://github.com/tidyverse/tidyverse" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@ldecicco-USGS & @jeffreyhanson, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @karthik know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="http://github.com/tidyverse/tidyverse">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@hadley) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

**If I understand correctly, my role is to perform these checks for each of the twenty tidyverse packages and also the _tidyverse R_ package. Therefore I have provided information for each item on the checklist below for each package. Please let me know if I have misunderstood anything.**

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="http://github.com/tidyverse/tidyverse">repository url</a>?
  - [x] tidyverse
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
    - [x] broom
    - [x] dplyr
    - [x] feather
    - [x] forcats
    - [x] ggplot2
    - [x] haven
    - [x] hms
    - [x] httr
    - [x] jsonlite
    - [x] lubridate
    - [x] modelr
    - [x] purr
    - [x] readr
    - [x] rvest
    - [x] readxl
    - [x] stringr
    - [x] tibble
    - [x] tidyr
    - [x] tidyverse
    - [ ] xml2
- [x] **Contribution and authorship:** Has the submitting author (@hadley) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
    - [x] tidyverse

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
    - [x] broom
    - [x] dplyr
    - [x] feather
    - [x] forcats
    - [x] ggplot2
    - [x] haven
    - [x] hms
    - [x] httr
    - [x] jsonlite
    - [x] lubridate
    - [x] modelr
    - [x] purr
    - [x] readr
    - [x] rvest
    - [x] readxl
    - [x] stringr
    - [x] tibble
    - [x] tidyr
    - [x] tidyverse
    - [x] xml2
- [x] **Functionality:** Have the functional claims of the software been confirmed?
    - [x] broom
    - [x] dplyr
    - [x] feather
    - [x] forcats
    - [x] ggplot2
    - [x] haven
    - [x] hms
    - [x] httr
    - [x] jsonlite
    - [x] lubridate
    - [x] modelr
    - [x] purr
    - [x] readr
    - [x] rvest
    - [x] readxl
    - [x] stringr
    - [x] tibble
    - [x] tidyr
    - [x] tidyverse
    - [x] xml2
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)
    - [x] broom
    - [x] dplyr
    - [x] feather
    - [x] forcats
    - [x] ggplot2
    - [x] haven
    - [x] hms
    - [x] httr
    - [x] jsonlite
    - [x] lubridate
    - [x] modelr
    - [x] purr
    - [x] readr
    - [x] rvest
    - [x] readxl
    - [x] stringr
    - [x] tibble
    - [x] tidyr
    - [x] tidyverse
    - [x] xml2

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
    - [x] broom
    - [x] dplyr
    - [x] feather
    - [x] forcats
    - [x] ggplot2
    - [x] haven
    - [x] hms
    - [x] httr
    - [x] jsonlite
    - [x] lubridate
    - [x] modelr
    - [x] purr
    - [x] readr
    - [x] rvest
    - [x] readxl
    - [x] stringr
    - [x] tibble
    - [x] tidyr
    - [x] tidyverse
    - [x] xml2
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
    - [x] broom
    - [x] dplyr
    - [x] feather
    - [x] forcats
    - [x] ggplot2
    - [x] haven
    - [x] hms
    - [x] httr
    - [x] jsonlite
    - [x] lubridate
    - [x] modelr
    - [x] purr
    - [x] readr
    - [x] rvest
    - [x] readxl
    - [x] stringr
    - [x] tibble
    - [x] tidyr
    - [x] tidyverse
    - [x] xml2
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
    - [x] broom
    - [x] dplyr
    - [ ] feather (documentation for `write_feather`---aliased as the documentation for `read_feather`---is missing an example that uses the `write_feather` function)
    - [x] forcats
    - [x] ggplot2
    - [x] haven
    - [x] hms
    - [x] httr
    - [x] jsonlite
    - [x] lubridate
    - [x] modelr
    - [x] purr
    - [x] readr
    - [x] rvest
    - [x] readxl
    - [x] stringr
    - [x] tibble
    - [x] tidyr
    - [x] tidyverse
    - [x] xml2
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
    - [x] broom
    - [x] dplyr
    - [x] feather
    - [x] forcats
    - [x] ggplot2
    - [x] haven
    - [x] hms
    - [x] httr
    - [x] jsonlite
    - [x] lubridate
    - [x] modelr
    - [x] purr
    - [x] readr
    - [x] rvest
    - [x] readxl
    - [x] stringr
    - [x] tibble
    - [x] tidyr
    - [x] tidyverse
    - [x] xml2
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
    - **Although all packages contain automated tests, some of the packages do not appear to have relatively high coverage (i.e. greater than 80%). This might be due to issues due with using continuous integration services (e.g. limited run time, disk space, or bandwidth). I have listed below some functions in tidyverse packages that might benefit from unit tests in case this is helpful.**
    - [x] broom
    - [x] dplyr
    - [x] feather (coverage not reported but contains tests for main functions)
    - [x] forcats
    - [x] ggplot2
    - [x] haven
    - [x] hms
    - [x] httr ([48% coverage coverage](https://codecov.io/gh/r-lib/httr?branch=master)): it might be useful to add tests for the functions: `cache_info`, `rerequest`, `httr_options`, `curl_docs`, `set_config`, `reset_config`, `with_config`, `cookies`, `set_cookies`, `httr_dr`, `set_envvar`, `get_envvar`, `BROWSE`, `oauth_endpoints` (endpoints for linkedin, twitter, vimeo, facebook, github, and azure), `progress`, `use_proxy`, `http_type`, `user_agent`, `verbose`, `write_disk`, `write_memory`, `write_stream`.
    - [x] jsonlite ([63% coverage coverage](https://codecov.io/github/jeroen/jsonlite?branch=master)): it might be useful to add tests for the functions: `stream_in`, `stream_out` (also verifies internal `apply_by_pages` and `stream_out_page` functions), `serializeJSON` (also verifies `pack` and `fixNativeSymbol`), `flatten`, `loadpkg`, `rbind_pages`, `unbox`.
    - [x] lubridate ([72% coverage coverage](https://codecov.io/gh/tidyverse/lubridate)): it might be useful to add tests for the functions: `fit_to_timeline`, `pretty_dates` (with seconds, minutes, hours, and months for coverage of internal `pretty_sec`, `pretty_min`, `pretty_hour`, `pretty_month` functions), `semester`, `dst`, `Date`.
    - [x] modelr
    - [x] purr
    - [x] readr
    - [x] rvest ([43% coverage](https://codecov.io/gh/tidyverse/rvest?branch=master)): it might useful to add tests for functions for forms (`set_values`, `submit_form`, `google_form`), sessions (`html_session`, `jump_to`, `follow_link`, `session_history`, `back`, `*.session` methods), and encoding (`guess_encoding`, `repair_encoding`).
    - [x] readxl
    - [x] stringr
    - [x] tibble
    - [x] tidyr
    - [x] tidyverse [58% coverage](https://codecov.io/gh/tidyverse/tidyverse?branch=master): perhaps it might be useful to add tests for: `tidyverse_deps`, `tidyverse_logo`.
    - [x] xml2
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support
    **I have checked the CODE_OF_CONDUCT, DESCRIPTION, README, .github/*.md and vignette files for these three pieces of information, and listed below which information might be missing for each tidyverse package. The left check box corresponds to (1), the middle check box to (2) and the third check box to (3). Please let me know if I have missed anything.**
    - [x] [x] [x] broom
    - [x] [x] [x] dplyr
    - [ ] [x] [ ] feather
    - [x] [x] [x] forcats
    - [x] [x] [x] ggplot2
    - [x] [x] [x] haven
    - [x] [x] [x] hms
    - [ ] [x] [ ] httr
    - [ ] [x] [ ] jsonlite
    - [x] [x] [x] lubridate
    - [x] [x] [x] modelr
    - [x] [x] [x] purr
    - [x] [x] [x] readr
    - [x] [x] [x] rvest
    - [x] [x] [x] readxl
    - [x] [x] [x] stringr (information available via files in tidyverse/.github when opening new issue)
    - [x] [x] [x] tibble
    - [x] [x] [x] tidyr
    - [x] [x] [x] tidyverse
    - [ ] [x] [ ] xml2

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by jeffreyhanson** [comment]
@hadley, I'm very sorry if this is a silly question and I don't want to waste your time, but I just wanted to double check if the pdf version of the manuscript linked above has the correct version of your manuscript? I was noticed that the `paper.Rmd` file has been updated more recently than the `paper.md` file in the repository which (if I understand correctly) is being used to generate the pdf version of the manuscript (e.g. line 117 of the documents are different).

**Comment by hadley** [comment]
@jeffreyhanson oh shoot, somehow we forgot to update the `.md`. Let me try and figure out what's gone wrong.

**Comment by hadley** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01686/joss.01686/10.21105.joss.01686.pdf)

**Comment by hadley** [comment]
Rendered preview is definitely ok now — sorry for the mess up.

**Comment by jeffreyhanson** [comment]
No worries, thanks for fixing it.

**Comment by jeffreyhanson** [comment]
The manuscript "Welcome to the tidyverse" provides an excellent overview of the tidyverse---an ecosystem of _R_ packages that share a vision for data science---and the _tidyverse R_ package---an _R_ package that helps manage this ecosystem---that will serve as a useful introduction for new data scientists. Currently, the tidyverse collection of packages spans eight core packages and twelve non-core packages. The core packages provide tools for importing (_readr R_ package), organizing (_tibble R_ package), manipulating (_dplyr_, _forcats_, _purr_, _stringr_, _tidyr R_ packages), and visualizing  data (_ggplot2 R_ package). The non-core packages provide more specialized functionalities, such as importing and exporting data from specific formats (e.g. _feather R_ package), manipulating specific data types (e.g. _lubridate R_ package), and converting complex and verbose model outputs into simple tabular formats (e.g. _broom R_ package). In my view, the manuscript is well-written for a broad audience. The tidyverse has had a profound influence over data science, and I look forward to seeing where it takes us in the future.

I have organized this review into three sections: **Major issues**, **Minor issues**, and **Assorted comments**. To be clear, the issues listed under **Assorted comments** are incredibly minor and do not substantially affect the quality of the manuscript. As a consequence, they can be ignored if the authors do not think changes are needed. Since the pdf version of the manuscript does not contain line numbers, I will refer to the line numbers as they appear in the [source rmarkdown document at the time of writing this review](https://github.com/tidyverse/tidyverse/blob/0271bf7f0a65870167fb3a897c4daa97905b31ba/vignettes/paper.Rmd).

@karthik, I have a few questions about exactly how some of the criteria in the **Reviewer checklist** should be applied to this particular submission (see the bold text in the checklist above). Once you've answered those questions, I can begin creating issues on the respective repositories if that's ok?

### Major issues

- Lines 115--131: In my opinion, the _tidyverse R_ package provides two additional functionalities that are not discussed in the manuscript. Firstly, attaching the _tidyverse R_ package will ensure that tidyverse packages are attached in the correct order. This is not trivial for beginners (e.g. attaching the _plyr_ and _dplyr R_ packages in the incorrect order will cause issues). Secondly, attaching the _tidyverse R_ package will inform the user about namespace conflicts in a concise and informative manner. This is very useful when many tidyverse and non-tidyverse packages are attached. For instance, conflicts arise when the _tidyverse R_ package is attached following the _raster R_ package because both the _tidyr R_ package---a core tidyverse package---and the _raster R_ package contain `extract()` functions. I think the _tidyverse R_ package is exceptionally useful when evaluated on its standalone merits---that is, disregarding the usefulness of the other packages that comprise the tidyverse---and I think readers would benefit from knowing about these functionalities.

### Minor issues

- Line 109: Perhaps it is worth providing a citation for R? I realize that many readers will likely already be familiar with R to some extent, but perhaps this be useful for readers that are less familiar with R?
- Lines 109--113: This manuscript is about describing (i) the collection of packages referred to as the tidyverse (covered in the **Components** and **Philosophy** sections) and (ii) the _tidyverse R_ package (covered in the **Tidyverse package** section). The **Summary** section does a fantastic job of summarizing the tidyverse collection of packages but I feel like it could provide more information about the _tidyverse R_ package itself. Perhaps the purpose of the _tidyverse R_ package could be mentioned in the **Summary** section?
- Lines 123--129: Is the _tidyverse R_ package not a core tidyverse package? It is not listed as a core package in Line 129. Maybe it would be useful to explicitly state if the _tidyverse R_ package is a core package or not?
- Line 131: I wonder if it is worth listing all of the non-core packages in the tidyverse? I suggest this for two reasons. Firstly, it is noted that one of the main purposes for the _tidyverse R_ package is that it provides a convenient method to install all of the packages in the tidyverse. Since there over ten non-core packages listed on the tidyverse package's REAMDE file, I suspect that listing all of these packages here would help the reader appreciate this functionality. Secondly, if one of the main aims of this manuscript is to describe the tidyverse, then I think it would be useful to list all of the packages that comprise the tidyverse. Although additional packages may be added to the tidyverse in the future, I think this would helpful to the reader.
- Throughout manuscript: The **Reviewer Checklist** requires that "the authors describe how this software compares to other commonly-used packages". This manuscript does not make such comparisons. For instance, the tidyverse _readr R_ package provides functions for importing data similar to the _utils_ (`read.table()`) and _data.table R_ packages. If this manuscript were specifically about the _readr R_ package then I would expect such comparisons to be present, but I personally think that detailed comparisons for all tidyverse packages would distract from the main purpose of the manuscript. I raise this issue here because it is on the checklist.

### Assorted comments

- Line 111: ", not to solve every possible problem". Perhaps a semicolon or em-dash would be more appropriate here than a comma?
- Line 117: Perhaps "downloading and installing" might be more appropriate than "downloading", since the following _R_ code uses `install.packages`?
- Line 129: Perhaps "tidyverse version 1.2.0" to make it clear that this sentence is referring to a specific version of the _tidyverse R_ package?
- Line 129: Is the ordering of these packages significant? If not, perhaps these packages could be listed alphabetically?
- Line 139: Perhaps "application programming interface (API)" instead of "API" for clarity?
- Line 139: This is a bit nitpicky but I do not think that data are necessarily "stored" in an API? I would say that data are accessed through an API? Apologies if I'm just plain wrong.
- Line 139: My impression is that terms like "data frame" and "flat file" will be very useful for individuals with experience in programming _R_, _Python_, or _SQL_. If such individuals are the intended audience for this manuscript, then using such terminology is ideal. If a broader audience is intended, then perhaps using different terms (or clarifying these terms) will be useful (e.g. terms like "tabular" or "spreadsheet" may be more informative for a broader audience).
- Lines 139 and 141: It appears that the term "data" is being used to refer to a singular entity, though some would argue that the term "data" is a plural. I do not know the grammar rules to which this journal subscribes, so I thought I would raise this point just in case.
- Line 139: Perhaps it might be useful to provide the full name of the format that corresponds to these file extensions (e.g. tab-separated values for tsv)?
- Line 139: I don't want to wade into the debate regarding if data is a singular or plural term, but I think that it might be useful to be consistent throughout the manuscript. My interpretation of the sentence containing the phrase "[...] import data stored in other common formats or retrieve it directly [...]" uses the term "data" to refer to both a plural and a singular entity. Perhaps the sentence might be clearer if written as something like "[...] import data from other common formats or the web directly"?
- Line 139 and elsewhere: The manuscript contains many embedded links which mean that URLs are not shown in the text (e.g. [like this](https://google.com)). I am not familiar enough with the correct reference style for this journal, so I thought I would raise this point just in case.
- Line 141: Perhaps "core tidyr package provides" might be useful to be consistent with the rest of the manuscript (e.g. "the core readr package", line 139)?
- References: The reference for Bache & Wickham (2014) has a lower case "r" in the title when referring to the R programming language.

**Comment by Kevin-Mattheus-Moerman** [comment]
@karthik can you follow up with the authors on the status here?

**Comment by batpigandme** [comment]
@Kevin-Mattheus-Moerman, so sorry to all the reviewers for the delay. Edits in the works.

**Comment by karthik** [comment]
Thank you Mara!

**Comment by karthik** [comment]
> @karthik, I have a few questions about exactly how some of the criteria in the Reviewer checklist should be applied to this particular submission (see the bold text in the checklist above). Once you've answered those questions, I can begin creating issues on the respective repositories if that's ok?

Hi @jeffreyhanson Thanks so much for this exhaustive checkbox list. Please go ahead with opening issues in respective repos. I realize this is a huge undertaking as this is a large suite of packages.

**Comment by karthik** [comment]
@batpigandme No rush, but when possible can you give me a rough timeline for when I should check in with you on revisions?

**Comment by batpigandme** [comment]
Hi @karthik,
We have team work week this week, but it's high on our priority list, and we're aiming to finish up next week. 

Do you happen to have a particularly good example of what responding to feedback should look like? Do we need to resolve all the checkboxes. 

Thanks for your help, and your patience!

**Comment by karthik** [comment]
@jeffreyhanson I'm answering a few more questions here.

> This repository contains the source code for the tidyverse R package. It does not contain the source code for all the tidyverse packages. However, the README file in this repository contains links to other GitHub repositories which contain the tidyverse packages. Personally, I believe this approach is more sensible than using a single repository for all tidyverse packages, but this particular approach is technically not acceptable based on my (super literal) interpretation of this criteria. @karthik, how should I proceed?

No need to interpret this super literally. As long as individual packages contain source code for each package, we are good to check this off.

> Although not all tidyverse packages contain a LICENSE file, it is worth mentioning that all tidyverse packages contain a DESCRIPTION file which indicates the software license by name. @karthik, do all packages still require a LICENSE file?

Don't all R packages require a LICENSE file? http://r-pkgs.had.co.nz/description.html#license even when the license itself is mentioned in the README. Can you point me to ones that don't?

> I have checked whether the submitting author is listed as an author in the DESCRIPTION file of each tidyverse package. I did not differentiate based on the role (e.g. aut, cre, ctb). @karthik, since this particular submission pertains to a collection of packages is it acceptable if @hadley is not an author of every single package?

That's fine in this case. It's clear that the Tidyverse idea itself is from Hadley, and this collection represents core packages that are part of that philosophy. So in this case I'd say it's perfectly fine that Hadley is not listed as an author on all of the packages.

**Comment by karthik** [comment]
@batpigandme  Here are a couple of good examples of responses:

https://github.com/openjournals/joss-reviews/issues/964
https://github.com/openjournals/joss-reviews/issues/741

> Do we need to resolve all the checkboxes.

Yes. or provide a reasonable explanation for ones that cannot be resolved so that the reviews can respond.

**Comment by hadley** [comment]
@karthik (just responding quickly to license question): From R-exts:

> Whereas you should feel free to include a license file in your source distribution, please do not arrange to install yet another copy of the GNU COPYING or COPYING.LIB files but refer to the copies on https://www.R-project.org/Licenses/ and included in the R distribution (in directory share/licenses). Since files named LICENSE or LICENCE will be installed, do not use these names for standard license files. To include comments about the licensing rather than the body of a license, use a file named something like LICENSE.note.

We work around this in (e.g.) `use_gpl_license()` by including a `.Rbuildignore`d `LICENSE.md` which R CMD check don't see locally and isn't included in the package bundle (and still allows you to use `LICENSE` for the (e.g.) template fields needed for to make the MIT license compatible with CRAN's requirements.)

We include license files in new packages, but we haven't made a systematic effort to include them in all older packages, so this is a good prompt for us to do so.

**Comment by batpigandme** [comment]
Thank you so much, @jeffreyhanson and @ldecicco-USGS, for you thoughtful reviews, and, @karthik, for your guidance throughout this process. 
Your input has made the paper much stronger.

We've responded to your feedback piece by piece in three markdown documents:

* @ldecicco-USGS's [here](https://github.com/tidyverse/tidyverse/blob/master/vignettes/review-ldecicco-USGS.md),  
* @jeffreyhanson's [here](https://github.com/tidyverse/tidyverse/blob/master/vignettes/review-jeffreyhanson.md), and  
* edits for the review checklist [here](https://github.com/tidyverse/tidyverse/blob/master/vignettes/review-checklist.md).

Of course, if you prefer, we're happy to include the bodies of these documents as comments in this issue.

Thank you again for your time and patience, and please let us know if there's anything you think we've overlooked, our outstanding items that we need to address.

**Comment by ldecicco-USGS** [comment]
The responses look fantastic to me, and now I'm super excited to see the development of the Tidyverse design guide.... 👯 💯 🎉

**Comment by karthik** [comment]
Thanks @batpigandme and @hadley!

> Of course, if you prefer, we're happy to include the bodies of these documents as comments in this issue.

Since it's part of the version history of Tidyverse, I think it's fine where it is as this review is tied to the repo.

I'll wait for @jeffreyhanson and @ldecicco-USGS to respond to your comments. 
@ldecicco-USGS let me know when you are ready to sign off.

**Comment by ldecicco-USGS** [comment]
Consider me signed off. 👍

**Comment by jeffreyhanson** [comment]
I am delighted to review the revised version of the "Welcome to the tidyverse" manuscript. In my view, provides a wonderful introduction to the tidyverse and the description of the _tidyverse R_ package. It is exceptionally well-written for a broad audience. The authors have addressed all of my issues with the previous version of the manuscript. I could not find any major or minor issues with this work, and have very few improvements to suggest for the manuscript. I re-checked the unresolved JOSS checklist items I flagged in my initial review, and almost all of them have been addressed. I have listed the few remaining unresolved checklist items below. I apologise if these checklist items have actually been addressed and I was just looking in the wrong files.

I have organized this review into three sections: **Major issues**, **Minor issues**, **Assorted comments**, and **Remaining checklist items**. To be clear, the issues listed under **Assorted comments** are incredibly minor and do not substantially affect the quality of the manuscript. As a consequence, they can be ignored if the authors do not think changes are useful. The issues listed under **Remaining checklist items** pertain to the JOSS reviewer checklist.

### Major issues

- None.

### Minor issues

- None.

### Assorted comments

- Line 190--191: Perhaps it might be worth describing the pipe as an operator when introducing it here? For example, instead of "... the pipe, `%>%`, ...", perhaps "... the pipe operator,`%>%`, ..." might be more informative? On the other hand, the term "operator" may be too technical for a broad-audience? I raise this just in case this is helpful.

### Remaining checklist items

- The _xml2 R_ package does not appear to contain a license file.
- The _feather R_ package does not provide an example of the `write_feather` function (see [Rd file here](https://github.com/wesm/feather/blob/master/R/man/read_feather.Rd)). Given that this function represents half of the functionality provided by this package, I think it would be important to provide an example of this function. However, such example code may not be possible given CRAN policies regarding `\donttest` and `\dontcheck` blocks in example code.
- The _feather_, _httr_, and the _jsonlite R_ packages appear to be missing information describing how other people can contribute to the software and/or where they can find help with using the software (e.g. in README or CODE_OF_CONDUCT files). I'm sorry if this issue was addressed and I'm just wrong about this. I note that the repositories for these package are not under the tidyverse GitHub organization, and so making such changes may not be trivial.

**Comment by jeffreyhanson** [comment]
Sorry for not mentioning this in the previous post, I reviewed the revised version by looking at the Rmarkdown document in the tidyverse/tidyverse repository. As such, it might be worth using the JOSS automatic thingummywhatsit bot to make an updated pdf version of the manuscript to double check that the references appear correctly.

**Comment by batpigandme** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #1686 with the following error: 

 /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-74bc29a6a731/lib/whedon.rb:190:in `block in parse_authors': Author (Evan Miller) is missing affiliation (RuntimeError)
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-74bc29a6a731/lib/whedon.rb:188:in `each'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-74bc29a6a731/lib/whedon.rb:188:in `parse_authors'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-74bc29a6a731/lib/whedon.rb:91:in `initialize'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-74bc29a6a731/lib/whedon/processor.rb:36:in `new'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-74bc29a6a731/lib/whedon/processor.rb:36:in `set_paper'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-74bc29a6a731/bin/whedon:55:in `prepare'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/base.rb:466:in `start'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-74bc29a6a731/bin/whedon:116:in `<top (required)>'
	from /app/vendor/bundle/ruby/2.4.0/bin/whedon:23:in `load'
	from /app/vendor/bundle/ruby/2.4.0/bin/whedon:23:in `<main>'

**Comment by batpigandme** [comment]
^^ whoops, premature PDF generation. Sorry about that!

**Comment by batpigandme** [comment]
Thanks again for your comments!

The xml2, httr, jsonlite, and feather packages have a somewhat murky relationship to the tidyverse. These packages are mostly installed because we expect them to be used pretty frequently but are owned by other organisations. We'll work to get finish the checklist for these packages (since it's good practice regardless of this article), but in the interests of unblocking this article, are we ok to publish without them?

**Comment by hadley** [comment]
It looks like compilation failed because an author doesn't have an affiliation? Is this required? We asked all authors to supply their preferred affiliation, and this author preferred not to have one.

**Comment by arfon** [comment]
> It looks like compilation failed because an author doesn't have an affiliation? Is this required? We asked all authors to supply their preferred affiliation, and this author preferred not to have one.

Unfortunately this is a limitation of our current template. You can solve this by creating an affiliation of 'None' and associating it with the author like this: https://github.com/tidyverse/tidyverse/pull/205

**Comment by arfon** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01686/joss.01686/10.21105.joss.01686.pdf)

**Comment by hadley** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01686/joss.01686/10.21105.joss.01686.pdf)

**Comment by hadley** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01686/joss.01686/10.21105.joss.01686.pdf)

**Comment by hadley** [comment]
Sorry for all the rebuilds — just trying to get the last details of the code and images perfect

@whedon generate pdf

**Comment by hadley** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01686/joss.01686/10.21105.joss.01686.pdf)

**Comment by hadley** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01686/joss.01686/10.21105.joss.01686.pdf)

**Comment by karthik** [comment]
@hadley You can also use http://whedon.theoj.org/ to compile the paper

**Comment by hadley** [comment]
Good to know now that I've finished all my tweaking 😬

**Comment by karthik** [comment]
There is always more tweaking to do before we head off to the DOI factory.

**Comment by arfon** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01686/joss.01686/10.21105.joss.01686.pdf)

**Comment by jeffreyhanson** [comment]
Yeah, I don't think the remaining checklist items should hold up publication. I'll sign off now - absolutely awesome work!

**Comment by karthik** [comment]
@hadley Can you please provide the latest version number for publication?

**Comment by karthik** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Attempting to check references...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- None

MISSING DOIs

- https://doi.org/10.1007/0-387-28695-0 may be missing for title: The Grammar of Graphics

INVALID DOIs

- None
```

**Comment by hadley** [comment]
@karthik what do people usually do with R packages? I'd like for the next release of tidyverse to an include a citation to this paper, but that introduces a chicken and egg problem.

I'll add the missing DOI momentarily.

**Comment by karthik** [comment]
I sent a PR with the missing DOI. 

> what do people usually do with R packages? I'd like for the next release of tidyverse to an include a citation to this paper, but that introduces a chicken and egg problem.

We're aware of this chicken and egg problem. For now, once the reviewers sign off, you archive the reviewed version on Zenodo, we add that DOI to your JOSS paper and then publish it. You can then add the JOSS citation to Tidyverse immediately and push a new release to GitHub/CRAN. That way people will see this paper in `citation()`.

JOSS does not publish a new paper with every version change, so you can still use this paper as a canonical reference. 

If you really want to add the citation _before_ you archive, it is:

```
Wickham et al., (2019). Welcome to the Tidyverse. Journal of Open Source Software, 4(43), 1686, https://doi.org/10.21105/joss.01686
```

**Comment by karthik** [comment]
Note: That DOI will not resolve for now because your paper isn't published yet.

**Comment by karthik** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Attempting to check references...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- None

MISSING DOIs

- https://doi.org/10.1007/0-387-28695-0 may be missing for title: The Grammar of Graphics

INVALID DOIs

- None
```

**Comment by hadley** [comment]
Ok, in that case can we use 1.3.0 for the version number? There are [no significant changes](https://github.com/tidyverse/tidyverse/blob/master/NEWS.md#tidyverse-1219000) between 1.2.1 and 1.3.0 except that 1.3.0 will include a citation to this paper.

Can you tell what is wrong with my DOI at https://github.com/tidyverse/tidyverse/blob/master/vignettes/paper.bib#L38 ?

**Comment by arfon** [comment]
> Can you tell what is wrong with my DOI at https://github.com/tidyverse/tidyverse/blob/master/vignettes/paper.bib#L38 ?

I wonder if Whedon is fussy about case? Can you try updating the BibTeX key to `doi` rather than `Doi`?

**Comment by karthik** [comment]
The trailing , could be the issue

**Comment by hadley** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Attempting to check references...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- None

MISSING DOIs

- https://doi.org/10.1007/0-387-28695-0 may be missing for title: The Grammar of Graphics

INVALID DOIs

- None
```

**Comment by hadley** [comment]
Tried both lower case and removing the `,` 😞

**Comment by arfon** [comment]
OK, ignore @whedon then about this. I'll file this as a bug.

**Comment by karthik** [comment]
@whedon set v1.3.0 as version

**Comment by whedon** [comment]
OK. v1.3.0 is the version.

**Comment by karthik** [comment]
@hadley This would be a good time to archive on Zenodo and share the Zenodo DOI here. Since you have so many authors you should make sure that Zenodo lists them in order before finishing that step.

**Comment by hadley** [comment]
Is there any specific zenodo advice for JOSS or is it obvious what I should be doing? 

One complexity is that the authors of the JOSS paper are the contributing authors across the entire tidyverse, as opposed to the single author (me) of the tidyverse package.

**Comment by hadley** [comment]
Or in the interests of simplicity, should I just add the authors of this paper to the tidyverse package?

**Comment by karthik** [comment]
Just the authors of this paper.

**Comment by hadley** [comment]
Working on it now. While doing so I noticed that the citation you gave me is slightly off — it should be "Welcome to the tidyverse" not "Welcome to the Tidyverse"

**Comment by hadley** [comment]
I have reserved 10.5281/zenodo.3547813, and will upload once I have re-assured myself this will definitely be the same as the version that appears on CRAN.

**Comment by hadley** [comment]
Ok, published at https://zenodo.org/record/3547813#.XdV2nC3MzUI

**Comment by karthik** [comment]
@whedon set 10.5281/zenodo.3547813, as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3547813" target="_blank">10.5281/zenodo.3547813</a> is the archive.

**Comment by karthik** [comment]
This is ready to publish. Over you to @openjournals/joss-eics

**Comment by kthyng** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1121

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1121, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- 10.1007/0-387-28695-0 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by kthyng** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1197622091178221570) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1122
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.01686](https://doi.org/10.21105/joss.01686)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by kthyng** [comment]
We're just waiting now for the DOI to resolve, should be any minute. In the meantime, congrats to @hadley and the other authors on your new paper!! Thanks to @karthik for editing and for the time and expertise of our reviewers, @ldecicco-USGS and @jeffreyhanson. 🎉 🎉

**Comment by kthyng** [comment]
And it's all there now! Congrats again!

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.01686/status.svg)](https://doi.org/10.21105/joss.01686)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.01686">
  <img src="https://joss.theoj.org/papers/10.21105/joss.01686/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.01686/status.svg
   :target: https://doi.org/10.21105/joss.01686
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.01686/status.svg)](https://doi.org/10.21105/joss.01686)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss

**Comment by karthik** [comment]
Congrats Tidyverse team! 🎉 🎈

https://joss.theoj.org/papers/10.21105/joss.01686

**Comment by hadley** [comment]
Thanks everyone — we greatly appreciate all your hard work making this paper happen!
